### PR TITLE
[MODEL-24537] sandbox: return at the result marker, not at terminal status

### PR DESCRIPTION
**Pull request by j1z0**
## What

`DataRobotWorkloadSandbox.run()` was strictly sequential: `_poll()` waited for a terminal workload status, and only then did `_fetch_logs()` read the `__DR_SANDBOX_RESULT__:` marker. The marker — the documented source of truth — is available long before terminal status.

Logs and status are now watched **concurrently**, and `run()` returns as soon as the marker is available. **The status poll is not removed.** It is still what fails a container that never starts, and it is the only source of a terminal `timeout` status or the `statusDetails.logTail` fallback. It just stopped being the gate.

## Why the gap exists

The runner is a **one-shot job scheduled as a long-running service**, because the workload-api only schedules `service` (or `nim`) artifacts. So:

1. The container runs the snippet in well under a second, prints its marker, and exits 0.
2. Kubernetes treats an exit-0 service container as a crash and restarts it — `Running (recovered from termination) last_reason=Completed last_exit_code=0`.
3. It runs again and emits the marker again.
4. Only after repeated restarts does k8s declare `CrashLoopBackOff` (logs show `back-off 10s`, then `back-off 20s`), and only then does workload-api report `errored`.

Gating the return on terminal status meant **waiting for Kubernetes to give up on a container that had already done its job.**

## Motivation — measured user impact

Ihar Baravy measured this on the BPA demo flow. Sandbox-backed tools only; non-sandbox tools were unchanged within ±2s:

| Tool | Before | After regression | Calls | Added |
|---|---|---|---|---|
| `transform_panel_dataset` | 14.0s | 33.0s | 18 | +308s |
| `create_chart_panel` | 11.0s | 39.3s | 6 | +170s |

## Measurement

Instrumented probe against **real staging** (`ENABLE_MCP_SANDBOX`), polling status and logs on independent clocks — the same method used to characterise the problem. 7 sandbox executions across 4 snippet shapes:

| snippet | submitted | first_log | MARKER | TERMINAL | saving | status | exitCode | markers |
|---|---|---|---|---|---|---|---|---|
| normal | 1.96s | 13.8s | 13.8s | 30.66s | 16.9s | errored | `null` | 2 |
| normal | 0.86s | 14.91s | 14.91s | 14.45s | 0s | errored | `null` | 2 |
| normal | 0.78s | 8.93s | 18.62s | 15.92s | 0s | errored | `null` | 2 |
| normal | 1.02s | 6.44s | 16.46s | 28.78s | 12.3s | errored | `null` | 3 |
| raises | 1.07s | 11.57s | 11.57s | 31.35s | 19.8s | errored | `null` | 3 |
| returns None | 0.93s | 8.52s | 18.7s | 33.86s | 15.2s | errored | `null` | 3 |
| times out | 0.78s | 14.21s | 16.55s | 48.23s | 31.7s | errored | `null` | 3 |
| **mean** | | | **15.8s** | **29.0s** | **13.7s** | | | |

`saving` is `max(0, TERMINAL − MARKER)` — what the old code spent that the new code does not. **46% of the mean call.** Two runs saw terminal status arrive slightly *before* the marker; those are a wash, not a regression, and they are included in the mean rather than discarded.

### What I could not verify

**I could not re-run the end-to-end `run()` timing from this branch against staging.** The machine slept mid-session and GlobalProtect now requires interactive sign-in (three `vpn.sh up` attempts failed with `tunnel did not carry traffic`; the `10.13.x` tunnel address is present but carries no traffic, and an authenticated `GET /account/info/` times out). So the table above measures the **before** state only — it is the marker-vs-terminal characterisation, not a stopwatch on the new code path against staging.

That said, `MARKER` is precisely the instant the new code returns at, and the new code path is covered by 16 new unit tests. But the honest statement is: the *saving* column is derived from measured staging timings, not observed end-to-end on this branch. Worth re-running once the VPN is back.

## Hazards

**1. `exit_code` comes from the terminal workload record.**
Measurement settled this: the workload record reports **`exitCode: null` for all 7 runs**, including the deliberately timing-out one — it describes a *service*, not a job. So `int(terminal.get("exitCode", 0) or 0)` was **already always 0** here, and a marker-first return loses nothing. `SandboxResult.exit_code` reports `0`.

That same finding means **the `exitCode == 124` check could never have fired on this backend** — a user-code timeout was silently returning as a successful `None`. Timeout detection now also keys on the runner's own `sandbox exceeded timeout of Ns` line (`protocol.has_timeout_marker`), which is in-band and works with no terminal record. The exit-code check is kept for backends/API versions that do populate it.

Two details confirmed by measurement, not assumed:
- The sentinel is checked across **stdout AND stderr**. Staging tags the runner's own writes to fd 2 as **INFO**, so the line arrives on the *stdout* side; only the platform's own `lrs-*` records come through at ERROR severity. A stderr-only check would have missed it entirely.
- The runner's timeout marker is `__DR_SANDBOX_RESULT__:null` — byte-identical to a snippet returning nothing. A marker alone is never proof of success, so a **null payload with no sentinel in hand and no terminal record** costs one short confirming re-read (`_NULL_MARKER_RECHECK_S`, 1s) before the run may succeed. Scoped to exactly that shape; a real return value pays nothing.

**2. Restarts emit multiple markers.**
Verified on staging: the logs endpoint returns records **newest-first** (`True` on all 7 runs), and terminal-gated runs had **2–3 markers** to choose between. Since `_partition_log_entries` joins in endpoint order and `parse_result_marker` takes the *last* marker line, resolution lands on the **oldest** marker — the first container attempt.

This is safe rather than lucky: code and inputs are baked into the workload at create time, so every restart re-runs the **identical** snippet — there is no earlier run with a different result to inherit. Returning at the first marker also strictly narrows the window in which duplicates can accumulate. Locked down by a test that feeds *deliberately different* payloads newest-first and asserts which one resolves, so a change in either ordering has to be deliberate.

**3. Teardown ordering.**
`_await_result()` fully completes before `run()`'s `finally`, so the marker is always read before the DELETE. The workload is still force-deleted on success, failure, timeout and cancellation. Asserted in the marker-first, timeout, never-starts, status-failure and cancellation tests.

**4. Genuine failures still fail.**
The `#638` behaviour is preserved exactly: **empty output has no stillness rule and polls to the deadline**, because it cannot be distinguished from "the container has not started yet" — a transient `ErrImagePull` errors the workload before the runner starts, k8s retries, and the container can begin a minute later. `_marker_deadline()` reproduces the old budgets with the terminal record now arriving mid-wait (not terminal → overall deadline; terminal failure → `max(deadline, terminal + flush)`; otherwise → `terminal + flush`).

One honest caveat: a **permanently** failing pull is indistinguishable from a retrying one from the client side, so that case still uses the provisioning budget. Shortening it would regress exactly the case `#638` exists for, so it is deliberately unchanged.

**5. Cancellation.**
The status task is named and always cancelled + drained in a `finally`, on every exit path. `_watch_status` swallows every non-cancellation exception, so the only thing the drain can raise is the `CancelledError` we requested — and if an outer cancellation interrupts the drain, the task is already cancelled and finishes without an unretrieved exception. `run()` still re-raises `CancelledError` after cleanup. A test asserts **no `dr-sandbox-status-*` task is left pending** after cancelling `run()`.

## Also fixed along the way

- `SandboxError` / `SandboxTimeout` raised here now carry `exit_code` and `stderr`. `observability.classify_outcome` already read both and was receiving neither, so **every OOM was being classified as `crash`**.
- A failing status endpoint no longer fails a run whose snippet demonstrably completed: the error is recorded and only re-raised when no marker was found either.
- A terminal `timeout` status no longer spends the 30s flush budget hunting a marker that cannot change the outcome.

## Test coverage

69 pass in `tests/drtools/unit/core/sandbox/` (53 existing, unmodified, + 16 new). Most new tests mock status as permanently `running`/`provisioning` — deliberately, since a terminal-gated implementation would poll such a workload to its deadline and raise, so they only pass if the return is genuinely marker-gated.

- marker returns without waiting for terminal status, and the status poll still ran concurrently (`call_count > 1`)
- `exit_code == 0` on a marker-first return
- runner timeout detected in-band with **no** terminal record → `SandboxTimeout(exit_code=124)`
- the same detection when the sentinel lands on stderr instead of stdout
- null marker gets its confirming re-read, and that re-read can still catch a timeout
- a snippet that legitimately returns `None` still succeeds
- a non-null marker pays no re-read cost (`call_count == 1`); a terminal record skips it too
- duplicate markers from restarts resolve deterministically to the first attempt (differing payloads)
- a marker-first return sees only one marker
- empty output keeps polling while status is non-terminal (ErrImagePull retry, hazard 4)
- a container that never starts still raises
- status-poll failure is non-fatal with a marker, fatal without one
- terminal `timeout` status does not burn the flush budget
- cancellation leaves no pending status task (hazard 5)

Gates: `ruff check` clean, `ruff format --check` clean, `scripts/check_imports.py` clean, version bumped to `0.29.8` in `pyproject.toml` + `uv.lock` with a CHANGELOG entry. `e2e-tests/uv.lock` and `docs/uv.lock` both sit at `0.29.6` on main (they already lag) so they are untouched, matching the most recent version-bump commits.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

> [!NOTE]
> **Medium Risk**
> This rewrites the core workload sandbox completion path (concurrency, timeout detection, and error surfacing), which affects every container-backed tool run, though behavior is heavily regression-tested.
> 
> **Overview**
> **Workload sandbox runs now finish as soon as the `__DR_SANDBOX_RESULT__:` log marker appears**, instead of waiting for a terminal Kubernetes/workload-api status (often CrashLoopBackOff after a one-shot runner exits). **0.29.8** bumps the package version and documents the change.
> 
> `DataRobotWorkloadSandbox` **polls OTEL logs and workload status concurrently** (`_await_result`, `_watch_logs`, `_watch_status`). Status is still used for never-start failures, `timeout` status, and `logTail` fallback, but **a found result marker wins over status-endpoint errors**. Log polling starts at submit with capped backoff; marker wait deadlines mirror the old terminal-gated flush rules via `_marker_deadline`.
> 
> **User-code timeouts on this backend** are detected with **`has_timeout_marker`** on the runner’s `sandbox exceeded timeout of` line (stdout and stderr), because workload records often have **`exitCode: null`**. Ambiguous **`__DR_SANDBOX_RESULT__:null`** without a sentinel gets a **1s confirming re-read** before success. **`SandboxTimeout` / `SandboxError`** from this path now pass **`exit_code` and `stderr`** for observability classification. Terminal **`timeout`** status **stops** the marker wait early.
> 
> `protocol.py` adds **`TIMEOUT_LOG_MARKER`** and **`has_timeout_marker`**. **16 new unit tests** cover marker-first return, timeout/null-marker edge cases, ErrImagePull retry polling, status failures, and cancellation cleanup.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit f8e13c52b68ef34e72b40fafbf31e53cfc32f1ce. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: src/datarobot_genai/drtools/core/sandbox/workload.py [hunk 1]
```diff
@@ -463,7 +610,7 @@ async def _fetch_logs(
                 # (see the constant above) — it may be a pull still retrying.
                 break
             await asyncio.sleep(delay)
-            delay = min(delay * 2, 3.0)
+            delay = min(delay * 2, _LOG_POLL_MAX_DELAY_S)
```
[the remark this answers is not in the document]
Comment (j1z0) [reply src/datarobot_genai/drtools/core/sandbox/workload.py:645]: Confirmed and fixed in ef9b66f6.

You are right, and the code's own docstring gave it away — it claimed "the give-up rules are unchanged from the terminal-gated version". Those rules were only *safe* because they were terminal-gated. Once the watcher starts at submit, and given that the runner prints a startup line before user code, stdout goes non-empty immediately and then sits unchanged for exactly as long as the snippet runs quietly. So a quiet snippet got abandoned at 30s while its container was still working, and `timeout_s` stopped bounding user code the moment that first line appeared.

The stillness break now additionally requires `watch.terminal is not None`. Before terminal status, still output means "running quietly", not "finished without a marker", and the wait is bounded by `_marker_deadline` instead. Corrected the docstring too, since it was asserting the opposite.

Regression test: `test_quiet_snippet_is_not_abandoned_before_terminal_status` — status stays non-terminal across several polls while stdout is frozen at the startup line, with the quiet window shrunk to 0.01s so repetition alone would fire it; the marker only arrives after. It fails on the old code.

Worth noting what this shook loose: the test harness's `_noop_sleep` returned without suspending, so `_watch_logs` could spin and starve the concurrent `_watch_status` task — `watch.terminal` stayed `None` no matter what the mocked status endpoint returned. Test-harness artifact only (a real `asyncio.sleep(delay)` always suspends), but it made the existing quiet-window test pass for the wrong reason. It now awaits `sleep(0)`; side effect is the suite dropping 489s -> 158s.

File: src/datarobot_genai/drtools/core/sandbox/workload.py [hunk 2]
```diff
@@ -382,65 +490,104 @@ def _extract_log_tail(terminal: dict[str, Any]) -> str:
...
+            logger.exception("workload-api logs fetch raised; will retry / fall back to logTail")
+            return "", ""
+        if resp.status_code >= 400:
+            logger.warning(
+                "workload-api logs fetch failed: %s %s",
+                resp.status_code,
+                resp.text,
+            )
+            return "", ""
+        return self._partition_log_entries(resp.json().get("data") or [])
```
[the remark this answers is not in the document]
Comment (j1z0) [reply src/datarobot_genai/drtools/core/sandbox/workload.py]: Fixed in ef9b66f6. Correct — the docstring promised "never raises" but `resp.json()` and `_partition_log_entries` sat outside the try, so only transport errors were covered. Both are now inside it, logging and returning `("", "")` so a bad read is retried exactly like a transport error. That matters here specifically because this endpoint is the source of truth for the result marker: one malformed 200 would have failed an execution that had already succeeded.

Regression test: `test_non_json_logs_response_does_not_fail_the_run` — a 200 with an HTML body, then a 200 with `data` as a string, then the real marker; the run returns its value.

File: src/datarobot_genai/drtools/core/sandbox/workload.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (j1z0) [reply src/datarobot_genai/drtools/core/sandbox/workload.py:645]: Correction to my previous reply: I said the regression test "fails on the old code." I then actually checked, and it did not — it **passed** pre-fix, so it was not covering this finding at all. It shrank the quiet window to 0.01s, but 0.01s never elapses between mocked polls, so the stillness branch was never reached from either direction. Vacuous.

Rewritten in 7efcf84c: window is `0.0` (stillness provably reached on the very next poll) and status **never** goes terminal, which makes the terminal gate the only thing that can keep the wait alive. Verified in both directions this time — passes on HEAD, and on HEAD~1 fails with:

```
SandboxTimeout: sandbox exceeded timeout waiting for a result from workload wkl_123
```

That is precisely the reported failure: a container still running a quiet snippet, abandoned at the quiet window and misreported to the caller as a timeout.

Also confirmed against real staging (single-container workload, `sum(range(1000))` / 45s-silent / 45s-chatty):

```
fast      16.18s  value=499500           PASS
quiet     60.30s  value=quiet-survived   PASS
chatty    59.99s  value=chatty-survived  PASS
```

The `quiet` shape is the one that matters — 45s of total silence, well past the 30s window — and it returns its real value rather than timing out.
